src/vs/workbench/services/textMate/browser/workerHost/textMateWorkerTokenizerController.ts

```diff
@@ -0,0 +1,155 @@
+		}));
+
+		const languageId = this._model.getLanguageId();
+		const encodedLanguageId = this._languageIdCodec.encodeLanguageId(languageId);
+		this._worker.acceptNewModel({
+			uri: this._model.uri,
+			versionId: this._model.getVersionId(),
+			lines: this._model.getLinesContent(),
+			EOL: this._model.getEOL(),
+			languageId,
+			encodedLanguageId,
+		});
+	}
+
+	public override dispose(): void {
+		super.dispose();
+		this._worker.acceptRemovedModel(this._model.uri.toString());
+	}
+
+	/**
+	 * This method is called from the worker through the worker host.
+	 */
+	public setTokensAndStates(versionId: number, rawTokens: ArrayBuffer, stateDeltas: StateDeltas[]): void {
+		// _states state, change{k}, ..., change{versionId}, state delta base, change{j}, ..., change{m}, current renderer state
+		//                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                    ^^^^^^^^^^^^^^^^^^^^^^^^^
+		//                | past changes                                       | future states
+
+		// Apply past changes to _states
+		while (
+			this._pendingChanges.length > 0 &&
+			this._pendingChanges[0].versionId <= versionId
+		) {
+			const change = this._pendingChanges.shift()!;
+			const op = lineArrayEditFromModelContentChange(change.changes);
+			op.applyTo(this._states);
+		}
+
+		const curToFutureTransformer1 = MonotonousIndexTransformer.fromMany(
+			this._pendingChanges.map((c) => lineArrayEditFromModelContentChange(c.changes))
+		);
+
+		const tokens = ContiguousMultilineTokensBuilder.deserialize(
+			new Uint8Array(rawTokens)
+		);
+
+		// Apply future changes to tokens
+		for (const change of this._pendingChanges) {
+			for (const innerChanges of change.changes) {
+				for (let j = 0; j < tokens.length; j++) {
+					tokens[j].applyEdit(innerChanges.range, innerChanges.text);
+				}
+			}
+		}
+
+		// Filter tokens in lines that got changed in the future to prevent flickering
+		// These tokens are recomputed anyway.
+		const b = new ContiguousMultilineTokensBuilder();
+		for (const t of tokens) {
+			for (let i = t.startLineNumber; i <= t.endLineNumber; i++) {
+				const result = curToFutureTransformer1.transform(i - 1);
+				if (result !== undefined) {
+					b.add(i, t.getLineTokens(i) as Uint32Array);
+				}
+			}
+		}
+		this._backgroundTokenizationStore.setTokens(b.finalize());
+
+		const curToFutureTransformer = MonotonousIndexTransformer.fromMany(
+			this._pendingChanges.map((c) => lineArrayEditFromModelContentChange(c.changes))
+		);
+
+		// Apply state deltas to _states and _backgroundTokenizationStore
+		for (const d of stateDeltas) {
+			let prevState = d.startLineNumber <= 1 ? this._initialState : this._states.get(d.startLineNumber - 1 - 1);
+			for (let i = 0; i < d.stateDeltas.length; i++) {
+				const delta = d.stateDeltas[i];
+				const state = applyStateStackDiff(prevState, delta)!;
+				this._states.set(d.startLineNumber + i - 1, state);
+
+				const offset = curToFutureTransformer.transform(d.startLineNumber + i - 1);
```

Comment:
There's one aspect I don't quite like performance wise here. These states are all for consecutive lines, yet we transform each line independently against the edits. It would be great if we could transform a range of lines against the pending edits (and the result would be possibly multiple ranges of lines, depending if the edits end up splitting the range or not).